### PR TITLE
Fixed #18012 -- Propagated reverse foreign keys from proxy models to their concrete class.

### PR DESCRIPTION
**Pull request by charettes**

### CODE REVIEW

File: django/db/models/fields/related.py [hunk 1]
```diff
@@ -681,7 +681,7 @@ def contribute_to_related_class(self, cls, related):
         # Internal FK's - i.e., those with a related name ending with '+' -
         # and swapped models don't get a related descriptor.
         if not self.remote_field.is_hidden() and not related.related_model._meta.swapped:
-            setattr(cls, related.get_accessor_name(), self.related_accessor_class(related))
+            setattr(cls._meta.concrete_model, related.get_accessor_name(), self.related_accessor_class(related))
```
Comment (charettes) [inline django/db/models/fields/related.py]: If we were to attach the descriptor on concrete model we might have to do adjustments in related field clash checks.

File: django/db/models/deletion.py [hunk 2]
```diff
@@ -53,18 +52,10 @@ def DO_NOTHING(collector, field, sub_objs, using):
 
 
 def get_candidate_relations_to_delete(opts):
-    # Collect models that contain candidate relations to delete. This may include
-    # relations coming from proxy models.
-    candidate_models = {opts}
-    candidate_models = candidate_models.union(opts.concrete_model._meta.proxied_children)
```
Comment (charettes) [inline django/db/models/deletion.py]: With this removal `proxied_children` is now unused in Django's code base.

**Comment by charettes** [comment]
Added the tests provided in #5404 to proof they are also passing with the proposed changes. Ideally we would not merge them since they are redundant with the single one added in 4b4ebd074911f8ff59f4e4c068279b4d9de0da8f and require the creation of multiple tables which slowdowns the test suite.

**Comment by charettes** [comment]
A last change I would like to propose is make `ProxyModel._meta.related_objects == ConcreteModel._meta.related_objects`. This is not the case actually because it breaks `get_fields(include_parents=True, include_hidden=True)` which ends up returning the same field twice. I'll see what I can do here.

File: django/db/models/deletion.py (hunk 2 above)
Comment (timgraham) [inline django/db/models/deletion.py]: Seems fine to remove to me.

**Comment by charettes** [comment]
I don't know why exactly but Github's UI seems to be confused, 32c116486ce441b433cb829d51fbc850d73d4df0 should be the last of the patch series. You can assert it is the case by looking at the [patch](https://github.com/django/django/pull/5378.patch) directly.

File: django/db/models/deletion.py (hunk 2 above)
Comment (charettes) [inline django/db/models/deletion.py]: Just did in c5dd36a0aec7b4ff249b3c9d2ae9e5d12a106bd7

**Comment by timgraham** [comment]
Looks good. How about docs?

**Comment by charettes** [comment]
Are you thinking about something else than release notes? A `versionchanged` somewhere in the `_meta` docs?

**Comment by timgraham** [comment]
Mostly release notes, unless you think a mention somewhere else is warranted.

File: django/db/models/fields/related_descriptors.py [hunk 3]
```diff
@@ -193,12 +193,13 @@ def __set__(self, instance, value):
...
         # If null=True, we can assign null here, but otherwise the value needs
         # to be an instance of the related class.
+        remote_model = self.field.remote_field.model
         if value is None and self.field.null is False:
             raise ValueError(
                 'Cannot assign None: "%s.%s" does not allow null values.' %
                 (instance._meta.object_name, self.field.name)
             )
-        elif value is not None and not isinstance(value, self.field.remote_field.model):
+        elif value is not None and not isinstance(value, (remote_model, remote_model._meta.concrete_model)):
```
Comment (akaariai) [inline django/db/models/fields/related_descriptors.py]: Is remote_model child of remote_model._meta.concrete_model if they are different?

Comment (charettes) [inline django/db/models/fields/related_descriptors.py]: Good catch, `isinstance(value, self.field.remote_field.model._meta.concrete_model)` should do in this case.

**Comment by charettes** [comment]
Will have a look at the remaining issues Sunday.

File: docs/releases/1.10.txt [hunk 4]
```diff
@@ -163,7 +163,11 @@ Migrations
 Models
 ^^^^^^
 
-* ...
+* Reverse foreign keys from proxy models are now propagated to their
+  concrete class. Before this change the reverse relation attached by a
```
Comment (timgraham) [inline docs/releases/1.10.txt]: The reverse relation .... is now accessible... and may be referenced in queryset filtering.

File: docs/releases/1.10.txt [hunk 5]
```diff
@@ -234,6 +238,15 @@ But it didn't prohibit nested non-relation fields as it does now::
     ...
     FieldError: Non-relational field given in select_related: 'name'
 
+``_meta.get_fields()`` returns consistent reverse fields for proxy models
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Before Django 1.10 the :meth:`~django.db.models.options.Options.get_fields`
+method returned different reverse fields when called on a proxy model compared
+to its proxied concrete class. This inconsistency was fixed by making sure the
```
Comment (timgraham) [inline docs/releases/1.10.txt]: by returning the ...

File: docs/releases/1.10.txt [hunk 6]
```diff
@@ -234,6 +238,15 @@ But it didn't prohibit nested non-relation fields as it does now::
     ...
     FieldError: Non-relational field given in select_related: 'name'
 
+``_meta.get_fields()`` returns consistent reverse fields for proxy models
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Before Django 1.10 the :meth:`~django.db.models.options.Options.get_fields`
```
Comment (timgraham) [inline docs/releases/1.10.txt]: 1.10,

File: docs/spelling_wordlist [hunk 7]
```diff
@@ -645,6 +645,7 @@ refactored
 refactoring
 refactorings
 refactors
+referenceable
```
Comment (timgraham) [inline docs/spelling_wordlist]: (chop when using suggested wording)

**Pull request merged**